### PR TITLE
[cmd/telemetrygen] Allow configuring Sum and Gauge metrics via CLI

### PR DESCRIPTION
**Pull request by marcelbirkner**
**Description:** <Describe what has changed.>
Adding a feature - Allow configuring Sum and Gauge metrics in `telemetrygen` via CLI

### CODE REVIEW

File: cmd/telemetrygen/internal/metrics/worker.go [hunk 1]
```diff
@@ -28,7 +29,13 @@ type worker struct {
 
 func (w worker) simulateMetrics(res *resource.Resource, exporter sdkmetric.Exporter) {
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/worker.go:30]: Instead of generating three metrics I would suggest having a CLI option that allows you to specify the metric type.

To keep things simple, let's do a first PR for adding support for adding sums first and we can tackle histograms in a separate PR.

File: cmd/telemetrygen/internal/metrics/worker.go [hunk 2]
```diff
@@ -28,7 +29,13 @@ type worker struct {
...
 	limiter := rate.NewLimiter(w.limitPerSecond, 1)
+
 	var i int64
+	value := 24.42
+	attrs := attribute.NewSet(attribute.KeyValue{
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/worker.go:35]: Let's also move the possibility of adding attributes to a separate PR

**Review by mx-psi — COMMENTED** [review]
Thanks for your PR! Let's split this into separate focused PRs so we can design this carefully :)

File: cmd/telemetrygen/internal/metrics/worker.go (hunk 1 above)
Comment (marcelbirkner) [reply cmd/telemetrygen/internal/metrics/worker.go:30]: Sounds good. I will split up the PRs today

File: cmd/telemetrygen/internal/metrics/worker.go (hunk 2 above)
Comment (marcelbirkner) [reply cmd/telemetrygen/internal/metrics/worker.go]: Removed attributes. Will create a separate PR

File: cmd/telemetrygen/internal/metrics/worker.go (hunk 1 above)
Comment (marcelbirkner) [reply cmd/telemetrygen/internal/metrics/worker.go:30]: Added new CLI parameter `metric-type`

image

File: cmd/telemetrygen/internal/metrics/worker.go (hunk 2 above)
Comment (mx-psi) [reply cmd/telemetrygen/internal/metrics/worker.go]: Thanks! Have a look at #26506, it may do everything you wanted

File: cmd/telemetrygen/internal/metrics/metrics_types.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package metrics
+
+import (
+	"errors"
+)
+
+type metricType string
+
+const (
+	metricTypeAll       metricType = "all"
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/metrics_types.go:10]: I think it would be better to design a more general mechanism to choose how many metrics to send and customize each one. This seems like something that would take some time to design, so my preference would be to remove the `all` option for now.

File: cmd/telemetrygen/internal/metrics/worker.go [hunk 4]
```diff
@@ -28,29 +29,63 @@ type worker struct {
...
 	for w.running.Load() {
-		rm := metricdata.ResourceMetrics{
-			Resource: res,
-			ScopeMetrics: []metricdata.ScopeMetrics{
-				{
-					Metrics: []metricdata.Metrics{
+		var metrics []metricdata.Metrics
+		if w.metricType == metricTypeGauge || w.metricType == metricTypeAll {
+			metrics = append(metrics, metricdata.Metrics{
+				Name: "gen.metric.gauge",
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/worker.go:38]: Let's keep the name as `"gen"` for all metrics (this implies removing the `"all"` option). I think we should keep the name as `"gen"` because:
1. It would be a breaking change for people already using this
2. We may want to add an option to change the name in the future and it makes sense to start with the same default always

File: cmd/telemetrygen/internal/metrics/worker.go [hunk 5]
```diff
@@ -28,29 +29,63 @@ type worker struct {
...
+				Data: metricdata.Gauge[int64]{
+					DataPoints: []metricdata.DataPoint[int64]{
+						{
+							StartTime: time.Now(),
+							Time:      time.Now().Add(1 * time.Second),
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/worker.go:43]: Let's leave this as it was, there is no need to change the gauge on this PR
```suggestion
							Time:      time.Now(),
```

File: cmd/telemetrygen/internal/metrics/worker.go [hunk 6]
```diff
@@ -28,29 +29,63 @@ type worker struct {
...
-							Data: metricdata.Gauge[int64]{
-								DataPoints: []metricdata.DataPoint[int64]{
-									{
-										Time:  time.Now(),
-										Value: i,
-									},
-								},
-							},
+							StartTime: time.Now(),
+							Time:      time.Now().Add(1 * time.Second),
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/worker.go:58]: This puts the point in the future, let's do it the other way around and have the `StartTime` be one second in the past
```suggestion
							StartTime: time.Now().Add(-1 * time.Second),
							Time:      time.Now(),
```

File: cmd/telemetrygen/internal/metrics/worker.go [hunk 7]
```diff
@@ -28,29 +29,63 @@ type worker struct {
...
+						},
+					},
+				},
+			})
+		} else if w.metricType == metricTypeSum || w.metricType == metricTypeAll {
+			metrics = append(metrics, metricdata.Metrics{
+				Name: "gen.metric.sum",
+				Data: metricdata.Sum[int64]{
+					IsMonotonic: true,
+					Temporality: metricdata.DeltaTemporality,
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/worker.go:54]: I think we should make this cumulative since this is the default for OpenTelemetry SDKs. @atoulme any thoughts on this?

File: cmd/telemetrygen/internal/metrics/worker.go [hunk 8]
```diff
@@ -28,29 +29,63 @@ type worker struct {
...
-							},
+							StartTime: time.Now(),
+							Time:      time.Now().Add(1 * time.Second),
+							Value:     i,
 						},
 					},
 				},
-			},
+			})
+		} else if w.metricType == metricTypeHistogram || w.metricType == metricTypeAll {
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/worker.go:64]: Let's leave histograms on a separate PR so we can think about what data to produce and how to make it meaningful

File: cmd/telemetrygen/internal/metrics/metrics_types.go (hunk 3 above)
Comment (marcelbirkner) [reply cmd/telemetrygen/internal/metrics/metrics_types.go]: Removed the "all" option

File: cmd/telemetrygen/internal/metrics/worker.go (hunk 4 above)
Comment (marcelbirkner) [reply cmd/telemetrygen/internal/metrics/worker.go]: Reverted metric name back to "gen"

File: cmd/telemetrygen/internal/metrics/metrics_types.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package metrics
```
Comment (mx-psi) [inline cmd/telemetrygen/internal/metrics/metrics_types.go:1]: Looks like you need to add the license header on this file per the lint:
```suggestion
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

package metrics
```

**Review by mx-psi — COMMENTED** [review]
Looking great :rocket: Let's wait a bit to see if @atoulme has any thoughts regarding the aggregation temporality, otherwise it's a matter of making CI pass and it will be ready to merge from my side

File: cmd/telemetrygen/internal/metrics/worker.go (hunk 7 above)
Comment (mx-psi) [reply cmd/telemetrygen/internal/metrics/worker.go:55]:
```suggestion
					Temporality: metricdata.CumulativeTemporality,
```

**Review by mx-psi — APPROVED** [review]

**Review by atoulme — APPROVED** [review]

**Comment by mx-psi** [comment]
Thank you @marcelbirkner :) merged
